File: app/src/main/java/com/richardluo/globalIconPack/reflect/Resources.java
package com.richardluo.globalIconPack.reflect;

import android.graphics.drawable.Drawable;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import java.lang.reflect.Method;

public class Resources {
  public static final Method getDrawableForDensity =
      XposedHelpers.findMethodExact(
          android.content.res.Resources.class,
          "getDrawableForDensity",
          int.class,
          int.class,
          android.content.res.Resources.Theme.class);

  public static Drawable getDrawable(
      android.content.res.Resources resources,
      int resId,
      android.content.res.Resources.Theme theme) {
    return getDrawableForDensity(resources, resId, 0, theme);
  }

  public static Drawable getDrawableForDensity(
      android.content.res.Resources resources,
      int resId,
      int iconDpi,
      android.content.res.Resources.Theme theme) {
    try {
      return (Drawable)
          XposedBridge.invokeOriginalMethod(
              getDrawableForDensity, resources, new Object[] {resId, iconDpi, theme});
    } catch (Exception e) {
      XposedBridge.log(e);
      return null;
    }
  }
}

File: app/src/main/kotlin/com/richardluo/globalIconPack/reflect/Resources.kt
package com.richardluo.globalIconPack.reflect

import android.content.res.Resources
import android.content.res.Resources.Theme
import android.graphics.drawable.Drawable
import com.richardluo.globalIconPack.utils.Logger.logE
import com.richardluo.globalIconPack.utils.getOrNull
import com.richardluo.globalIconPack.utils.method
import io.github.libxposed.api.XposedInterface
import java.lang.reflect.Method

object Resources {
  val getDrawableForDensityM: Method? by lazy {
    Resources::class
      .java
      .method(
        "getDrawableForDensity",
        Int::class.javaPrimitiveType,
        Int::class.javaPrimitiveType,
        Theme::class.java,
      )
  }

  context(xposed: XposedInterface)
  fun getDrawableForDensity(
    thisObj: Resources,
    resId: Int,
    iconDpi: Int,
    theme: Theme?,
  ): Drawable? {
    return runCatching {
      xposed
        .getInvoker(getDrawableForDensityM ?: return null)
        .setType(XposedInterface.Invoker.Type.ORIGIN)
        .invoke(
          thisObj,
          resId,
          iconDpi,
          theme,
        ) as? Drawable
    }
      .getOrNull { logE(it) }
  }
}
